File: 3.x.x/crates-utopia/cache/src/adapter/memory.rs
use std::collections::HashMap;

use parking_lot::Mutex;

use crate::adapter::Adapter;
use crate::error::CacheError;
use crate::value::{is_empty_key, unix_now, CacheValue, LoadResult, SaveResult};

struct Entry {
    time: i64,
    data: CacheValue,
}

/// PHP `Utopia\Cache\Adapter\Memory`.
#[derive(Default)]
pub struct Memory {
    store: Mutex<HashMap<String, Entry>>,
}

impl std::fmt::Debug for Memory {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("Memory")
            .field("len", &self.store.lock().len())
            .finish()
    }
}

impl Memory {
    #[must_use]
    pub fn new() -> Self {
        Self::default()
    }
}

impl Adapter for Memory {
    fn load(&self, key: &str, ttl: i64, _hash: &str) -> Result<LoadResult, CacheError> {
        if is_empty_key(key) {
            return Ok(LoadResult::Miss);
        }
        let store = self.store.lock();
        match store.get(key) {
            Some(saved) if saved.time + ttl > unix_now() => Ok(LoadResult::Hit(saved.data.clone())),
            _ => Ok(LoadResult::Miss),
        }
    }

    fn save(&self, key: &str, data: &CacheValue, _hash: &str) -> Result<SaveResult, CacheError> {
        if is_empty_key(key) || data.is_php_empty() {
            return Ok(SaveResult::Failed);
        }
        self.store.lock().insert(
            key.to_owned(),
            Entry {
                time: unix_now(),
                data: data.clone(),
            },
        );
        Ok(SaveResult::Saved(data.clone()))
    }

    fn touch(&self, key: &str, _hash: &str) -> Result<bool, CacheError> {
        if is_empty_key(key) {
            return Ok(false);
        }
        let mut store = self.store.lock();
        match store.get_mut(key) {
            Some(saved) => {
                saved.time = unix_now();
                Ok(true)
            }
            None => Ok(false),
        }
    }

    fn list(&self, _key: &str) -> Result<Vec<String>, CacheError> {
        Ok(Vec::new())
    }

    fn purge(&self, key: &str, _hash: &str) -> Result<bool, CacheError> {
        if is_empty_key(key) {
            return Ok(false);
        }
        Ok(self.store.lock().remove(key).is_some())
    }

    fn flush(&self) -> Result<bool, CacheError> {
        self.store.lock().clear();
        Ok(true)
    }

    fn ping(&self) -> bool {
        true
    }

    fn get_size(&self) -> Result<i64, CacheError> {
        Ok(self.store.lock().len() as i64)
    }

    fn get_name(&self, _key: Option<&str>) -> String {
        "memory".into()
    }
}

File: packages/cache/src/Adapter/Memory.php
<?php

namespace Utopia\Cache\Adapter;

use Utopia\Cache\Adapter;

class Memory implements Adapter
{
    /**
     * @var array<string, mixed>
     */
    public $store = [];

    /**
     * @param  int  $ttl time in seconds
     * @param  string  $hash optional
     */
    public function load(string $key, int $ttl, string $hash = ''): mixed
    {
        if ($key !== '' && $key !== '0' && isset($this->store[$key])) {
            /** @var array{time: int, data: string} */
            $saved = $this->store[$key];

            return ($saved['time'] + $ttl > time()) ? $saved['data'] : false; // return data if cache is valid
        }

        return false;
    }

    /**
     * @param  array<int|string, mixed>|string  $data
     * @param  string  $hash optional
     * @param  int  $ttl time in seconds
     * @return bool|string|array<int|string, mixed>
     */
    public function save(string $key, array|string $data, string $hash = '', int $ttl = 0): bool|string|array
    {
        if ($key === '' || $key === '0' || empty($data)) {
            return false;
        }

        $saved = [
            'time' => time(),
            'data' => $data,
        ];

        $this->store[$key] = $saved;

        return $data;
    }

    /**
     * @param  string  $hash optional
     */
    public function touch(string $key, string $hash = ''): bool
    {
        if ($key === '' || $key === '0' || ! isset($this->store[$key])) {
            return false;
        }

        /** @var array{time: int, data: string|array<int|string, mixed>} $saved */
        $saved = $this->store[$key];
        $saved['time'] = time();
        $this->store[$key] = $saved;

        return true;
    }

    /**
     * @return string[]
     */
    public function list(string $key): array
    {
        return [];
    }

    /**
     * @param  string  $hash optional
     */
    public function purge(string $key, string $hash = ''): bool
    {
        if ($key !== '' && $key !== '0' && isset($this->store[$key])) { // if a key is passed and it exists in cache
            unset($this->store[$key]);

            return true;
        }

        return false;
    }

    public function flush(): bool
    {
        $this->store = [];

        return true;
    }

    public function ping(): bool
    {
        return true;
    }

    /**
     * Returning total number of keys
     */
    public function getSize(): int
    {
        return \count($this->store);
    }

    public function getName(?string $key = null): string
    {
        return 'memory';
    }
}
